sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -2331,11 +2403,11 @@ runtime.
 
 === Update More Command Types
 
-Support updating arguments to types of nodes other that kernel execution
+Support updating arguments to types of nodes other than kernel execution
 commands.
 
-**UNRESOLVED** Should be added for at least memory copy nodes, however
-full scope of support needs to be designed and implemented.
+**UNRESOLVED** Should be added for at least memory copy nodes and host-tasks.
```

Comment:
Could you describe the intended usage of barriers in this case, and whether you are wanting them to be updatable themselves or just to be allowed in graphs which can be updated?

If using barriers to define edges in the graph then we can't really allow those dependencies to be updated as that would change the topology. If we were to remove the restrictions on events from outside the graph though then it could be allowed to update those.

Reply:
> Could you describe the intended usage of barriers in this case, and whether you are wanting them to be updatable themselves or just to be allowed in graphs which can be updated?

We use barriers for pretty much all dependencies except for those promised by in-order queues, both internal (intra-graph) and external (outgoing and incoming), to make the high-level scheduling code compatible with CUDA.

The internal edges are "static", so no changes to topology here. The external incoming edges are planned to be dependent  on the proposed `dynamic_event`s, which will get updated (the edge stays the same, only the external event gets changed). The external outgoing edges are static as well.

Reply:
OK thanks for the clarification. We might need an overload of `ext_oneapi_barrier` which takes `dynamic_event`s to support this case then.

Reply:
I've been discussing this internally with the team and I think it makes sense to move discussion around barriers accepting dynamic events to that PR.

When it comes to allowing barriers in graphs that are being recorded I think that is a fair ask, though I have some thoughts about it:

- The current approach is very restrictive, but is the safest since as with most things we cannot guarantee that we could throw errors in the right cases when comparing nodes which cannot be updated to check for changes.
- Allowing barriers as a single exception is probably undesirable since it is both unintuitive and doesn't look great in the spec.
- To my mind if we allow one type of node to be a no-op during whole-graph update we should allow all types, but this multiplies the potential for confusion if we can't diagnose errors. We can throw warnings that nodes will not be updated if they are not kernel nodes to help alleviate this but that's implementation defined rather than being a SYCL standard thing.

Tagging @gmlueck for thoughts on removing these restrictions for whole-graph update.

Reply:
Sorry, I'm not following.  Is the request to:

* Allow a recorded graph to contain a barriers (e.g. call to `ext_oneapi_barrier`), and
* Allow "whole graph update" for such a graph, where the "source" graph also has a matching barrier

What is the problem with supporting this?  How is this different from having a graph where one of the nodes has "event" dependencies?  In fact, isn't a barrier semantically identical to having an empty kernel with "event" dependencies?

Reply:
Yes that is the request. The main issue I think is that currently we explicitly only allow kernel nodes, and any other type is an error (including both empty and barrier nodes). Barrier nodes are effectively empty nodes as you say but we do distinguish them for the user with `node_type::ext_oneapi_barrier`. So we would likely need to allow both empty and barrier nodes.

It's not so much an issue of supporting it in the implementation, but more than I was concerned about it being unintuitive that nodes could be passed with changed parameters but just be skipped for update, especially given the problems already discussed in this PR around reliably throwing errors.

If that is not a big concern then I am definitely in favour of allowing them to improve usability for this use case, though I do think it raises the question that if we allow these types why not just allow all of them.

Reply:
> It's not so much an issue of supporting it in the implementation, but more than I was concerned about it being unintuitive that nodes could be passed with changed parameters but just be skipped for update, especially given the problems already discussed in this PR around reliably throwing errors.

Can you explain this more?  This is the part I don't understand.  What is being "skipped"?  If it's easier to have a call, we could do that.